x-pack/plugins/security_solution/server/lib/detection_engine/signals/build_bulk_body.test.ts

```diff
@@ -632,3 +645,341 @@ describe('buildSignalFromEvent', () => {
     expect(signal).toEqual(expected);
   });
 });
+
+describe('recursive intersection between objects', () => {
+  test('should treat numbers and strings as unequal', () => {
+    const a = {
+      field1: 1,
+      field2: 1,
+    };
+    const b = {
+      field1: 1,
+      field2: '1',
+    };
+    const intersection = objectPairIntersection(a, b);
+    const expected = {
+      field1: 1,
+    };
+    expect(intersection).toEqual(expected);
+  });
+
+  test('should strip unequal numbers and strings', () => {
+    const a = {
+      field1: 1,
+      field2: 1,
+      field3: 'abcd',
+      field4: 'abcd',
+    };
+    const b = {
+      field1: 1,
+      field2: 100,
+      field3: 'abcd',
+      field4: 'wxyz',
+    };
+    const intersection = objectPairIntersection(a, b);
+    const expected = {
+      field1: 1,
+      field3: 'abcd',
+    };
+    expect(intersection).toEqual(expected);
+  });
+
+  test('should handle null values', () => {
+    const a = {
+      field1: 1,
+      field2: '1',
+      field3: null,
+    };
+    const b = {
+      field1: null,
+      field2: null,
+      field3: null,
+    };
+    const intersection = objectPairIntersection(a, b);
+    const expected = {
+      field3: null,
+    };
+    expect(intersection).toEqual(expected);
+  });
+
+  test('should handle explicit undefined values and return undefined if left with only undefined fields', () => {
+    const a = {
+      field1: 1,
+      field2: '1',
+      field3: undefined,
+    };
+    const b = {
+      field1: undefined,
+      field2: undefined,
+      field3: undefined,
+    };
+    const intersection = objectPairIntersection(a, b);
+    const expected = undefined;
+    expect(intersection).toEqual(expected);
+  });
+
+  test('should strip arrays out regardless of whether they are equal', () => {
+    const a = {
+      array_field1: [1, 2],
+      array_field2: [1, 2],
+    };
+    const b = {
+      array_field1: [1, 2],
+      array_field2: [3, 4],
+    };
+    const intersection = objectPairIntersection(a, b);
+    const expected = undefined;
+    expect(intersection).toEqual(expected);
+  });
+
+  test('should strip fields that are not in both objects', () => {
+    const a = {
+      field1: 1,
+    };
+    const b = {
+      field2: 1,
+    };
+    const intersection = objectPairIntersection(a, b);
+    const expected = undefined;
+    expect(intersection).toEqual(expected);
+  });
+
+  test('should work on nested fields', () => {
```

Comment:
Nit, but I feel like `nested` is an overloaded term, so might want to consider something different... `should work on objects of objects`?